gitlab/client.py

```diff
@@ -75,7 +81,34 @@ def __init__(
         keep_base_url: bool = False,
     ) -> None:
 
+        # We only support v4 API at this time
+        if api_version not in ("4",):
+            raise ModuleNotFoundError(f"gitlab.v{api_version}.objects")
+        # NOTE: We must delay import of gitlab.v4.objects until now or
+        # otherwise it will cause circular import errors
+        from gitlab.v4 import objects
+
         self._api_version = str(api_version)
+
+        if http == "requests":
+            self.client = _RequestsClient(
+                url=url,
+                session=session,
+                retry_transient_errors=retry_transient_errors,
+                ssl_verify=ssl_verify,
+                timeout=timeout,
+                http_username=http_username,
+                http_password=http_password,
+                job_token=job_token,
+                oauth_token=oauth_token,
+                private_token=private_token,
+            )
+            self.session = self.client.get_session
+        elif http == "httpx":
+            from .clients._httpxclient import _HttpxClient
+
+            self.client = _HttpxClient(client=client)
```

Comment:
Is it necessary that we offload all of this to the backend clients? It mostly just sets up auth headers, so ideally we could keep most of the logic in the existing client.

I have a suspicion some of this is needed because of `HTTPBasicAuth` in `_set_auth_info()`. In fact, this is sort of dead code as GitLab has long disabled basic auth via username/password. So instead we could perhaps remove that part of the code. WDYT?

Reply:
I would suggest to break it into two steps. Right now to do the minimal change of moving the same code from the general client to the requests client and do some cleanup in a follow up PR.